app/Http/Middleware/VersionSwitcher.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace App\Http\Middleware;
+
+/**
+ * Router middleware that switches the controller registered to current route
+ * depending on "Accept-Version" header.
+ */
+class VersionSwitcher
+{
+    public function handle($request, $next)
+    {
+        // If no version has been specified either in header or config do nothing.
+        if (!$version = $request->header('Accept-Version') ?? config(('api.version'))) {
+            return $next($request);
+        }
+
+        $route = $request->route();
+
+        $request->setRouteResolver(function () use ($version, $route) {
+
+            return array_map(function ($routeComponent) use ($version) {
+                // Only handle routes that has a controller defined.
+                if ($uses = $routeComponent['uses'] ?? null) {
+                    // Only handle controller registrations
+                    // with the pattern: [class]@[method].
+                    if (preg_match('/^([^@]+)@(.*)$/', $uses, $m)) {
+                        list(, $classFrom, $method) = $m;
+
+                        // Squeeze in a \v[version]\ before the controller class name.
+                        $replacement = sprintf('$1v%d\\\\$2', $version);
+                        $classTo = preg_replace('/(.*\\\\)([\w]+)$/', $replacement, $classFrom);
```

Comment:
I have to fold here. I consider myself OK at regular expressions but I do not understand what is going on here 😵 .

It we simply want to prepend the version to the controller class name could we do like so:

```php
$classTo = sprintf('\v%d\%s', $version, $classFrom);
```

There might be other considerations regarding class name resolving that I am not aware of though.

Reply:
`$classFrom` is the full namespace to the class: `App\Http\Controllers\ListController` so you cannot prepend the version number like that.
The reason that is har to read is that if you in a regex need to have a single backslash you need four in total to escape it!
I'll keep it regexed for now although it looks like an old language spoken by black witches.